Add a new auto-update feature

Michael-Sinz opened the pull request:
Add a new auto-update feature that can automatically find all VMSS pools (or pick specific ones) and then find if there are any nodes that are worthy of being used as a new prototype.

This makes use of some node annotations to make this know if the node may have newer/better OS patches/etc on them.

I have yet to build the helm chart for this but wanted to get the basic code in such that we have containers with that code.

(Already tested this in my private environment)

The new command:  auto-update - this does not replace the update command and in fact uses the update command to do the actual work.  The update command has not been changed (other than some minor code refactorings)

Arguments the same as the original update except:
1) No --target-node

1) --target-vmss (optional) can list 1 or more VMSS pools to operate over.  The default will be all VMSS pools.

2) --pending-reboot-annotation is the name of the annotation that, if it exists, the node is to be considered pending a reboot and not be used

3) --last-patch-annotation is the name of the annotation that holds the timestamp of the last OS patch/needed reboot.  This is used to tell if this node has some potentially newer code.  The annotation format must have a "YYYY-MM-DD" in its value, such as 2020-12-24 - it can have a full timestamp but the date is the key/critical part

4) --node-ignore-annotations - optional and can be multiple - more annotations that if they exist will cause the node to be ignored from consideration

5) --minimum-ready-time - default 60 seconds - amount of time that a node must be ready before it is considered a potential target node.  The default is very small.  Values can include modifiers such as 12m for 12 minutes, 3h for 3 hours, or 2d for 2 days (along with either 30 or 30s for 30 seconds)

6) --minimum-candidates - default 1 - the minumum number of potential candidates before any candidate is selected.  This way, on larger clustes, one can say that at least 10 nodes must have reached the stable state with the latest updates before we find it reliable enough to use one as a prototype node

7) --ignore-nodes - default none - zero or more node names that should be ignored from consideration

8) --dry-run - if given, all of the logic will run but the actual update to the prototype will not be executed.

A future commit will include more documentation updates as to how this all works.

Michael-Sinz committed:
Add a new auto-update feature

Add a new auto-update feature that can automatically find all
VMSS pools (or pick specific ones) and then find if there are any
nodes that are worthy of being used as a new prototype.

This makes use of some node annotations to make this know if the
node may have newer/better OS patches/etc on them.

I have yet to build the helm chart for this but wanted to get the
basic code in such that we have containers with that code.

(Already tested this in my private environment)

The new command:  auto-update

Arguments the same as the original update except:
0) No more --target-node

1) --target-vmss (optional) can list 1 or more VMSS pools to operate over.  The default will be all VMSS pools.

2) --pending-reboot-annotation is the name of the annotation that, if it exists, the node is to be considered pending a reboot and not be used

3) --last-patch-annotation is the name of the annotation that holds the timestamp of the last OS patch/needed reboot.  This is used to tell if this node has some potentially newer code.  The annotation format must have a "YYYY-MM-DD" in its value, such as 2020-12-24 - it can have a full timestamp but the date is the key/critical part

4) --node-skip-annotations - optional and can be multiple - more annotations that if they exist will cause the node to be skipped from consideration

5) --minimum-ready-time - default 60 seconds - amount of time that a node must be ready before it is considered a potential target node.  The default is very small.  Values can include modifiers such as 12m for 12 minutes, 3h for 3 hours, or 2d for 2 days (along with either 30 or 30s for 30 seconds)

6) --minimum-candidates - default 1 - the minumum number of potential candidates before any candidate is selected.  This way, on larger clustes, one can say that at least 10 nodes must have reached the stable state with the latest updates before we find it reliable enough to use one as a prototype node

7) --running-on-node - default nothing - this will be provided by the helm chart to identify the node this code is executing on such that it eliminates that node from the potential candidate pool.  This is note needed if you are sure this is running on non-VMSS nodes but there are systems where the policy is that "user" containers do not run on master nodes.  It is a trivial feature for helm and the yaml in kubernetes to provide so might as well provide it in the helm chart.

A future commit will include more documentation updates as to how this
all works.

Michael-Sinz committed:
Make the NODE_ID fill in the default for --running-on-node

This uses the environment variable NODE_ID to fill in the
--running-on-node default value.  This is already set in our
helm charts and will simpify usage in most cases.

Michael-Sinz committed:
A bit more refactoring

jackfrancis commented on vmss-prototype/vmss-prototype:
```diff
@@ -851,13 +872,148 @@ def vmss_prototype_update(sub_args):
+                logging.debug('SKIP: Node {0} does not have a last patch annotation: {1}'.format(node_name, sub_args.last_patch_annotation))
+                continue
+
+            last_patch_date_match = re.search(r'\d\d\d\d-\d\d-\d\d', last_patch)
```
I would prefer if we standardize to RFC3339 for our time type to accommodate more fine-grained timestamps. If we want to be opinionated about only allowing a day's worth of resolution, we can easily convert RFC3339 to a lower-resolution format.

Michael-Sinz replied:
I was doing this such that our (Skyman) current annotations could work as our annotations include additional information plus the timezone offset syntax (such as "reason@2021-01-09T09:32:37+00:00")

Michael-Sinz replied:
Note that this whole process is just to produce a "version" string that can map with the shared image gallery version such that the version compare can work below.

jackfrancis replied:
I see that the regex match pattern is compatible w/ RFC3339, nothing to do here

jackfrancis commented:
/lgtm

jackfrancis commented on vmss-prototype/vmss-prototype:
```diff
@@ -1028,18 +1259,27 @@ def register_commands(subparser):
+                        help='The name of the annotation that, if it exists, the node is pending reboot')
+    parser.add_argument('--last-patch-annotation', required=True, type=str,
+                        help='The name of the annotation that holds the timestamp at which the last patch was applied')
+    parser.add_argument('--node-skip-annotation', action='append', nargs='+', required=False, type=str,
```
maybe `--node-ignore-annotation` instead? (very mild opinion here)

Michael-Sinz replied:
If we change this to ignore, should the log entries say IGNORE rather than SKIP?

My guess is that they should as it would be more consistent.

Michael-Sinz replied:
So, I think having things as IGNORE rather than SKIP is ok - still not a great log line...

Michael-Sinz replied:
Now I know what I did not like.  IGNORE is a command IGNORED is what happened.

Ready to merge now?

jackfrancis commented on vmss-prototype/vmss-prototype:
```diff
@@ -1028,18 +1259,27 @@ def register_commands(subparser):
+                        help='Minimum time a node must in ready state before it is considered a candidate (default %(default)s)')
+    parser.add_argument('--minimum-candidates', default=1, type=MinimumCandidates,
+                        help='Minimum number of acceptable candidates (default: %(default)s)')
+    parser.add_argument('--running-on-node', default=os.getenv('NODE_ID', None), type=str,
```
How about we just change this to an array of node names, if we ever want to statically "disallow" one or more nodes? The nominal case will be helm auto-populating this with one node.

In which case it might be `--ignore-nodes` (or `--skip-nodes` if we prefer skip over ignore)

Michael-Sinz replied:
Interesting - currently it is filled in with the NODE_ID from the environment that gets set up with the yaml.  Will have to look into changing into a multi-value.  Will also have to look how to best deal with them in helm command line (it is not the easiest to deal with lists of things)  Maybe the first step is to just make the command support it and then later deal with helm if we want to.

Michael-Sinz replied:
So, what I did was to *always* respect NODE_ID and to make --ignore-nodes to allow multiple node names to be listed (default none) that are ignored

Michael-Sinz committed:
Review feedback plus some other changes

Changed SKIP to IGNORE and matching argument names

Added a dry-run option
Added rendering of the status after it completes
Added ignoring of cordoned nodes
Changed the node to ingore to be a list
Changed to always ignore the node we are running on (no option)

Michael-Sinz committed:
The debug logs now say IGNORED rather than IGNORE

IGNORED is what happened, IGNORE is a command

jackfrancis commented:
changes lgtm
